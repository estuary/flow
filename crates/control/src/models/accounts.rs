use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

use crate::models::Id;

/// Accounts are the owners of resources and the actors performing actions
/// within the system. An Account may represent a human, a company, or an
/// automated process.
///
/// An `Account` is a Rust representation of the Postgres database understanding
/// of the account.
#[derive(Clone, Debug, Deserialize, FromRow, Serialize)]
pub struct Account {
    /// When this record was created.
    pub created_at: DateTime<Utc>,
    /// A freeform representation of the name. eg. "Wiley Coyote" or "Acme Anvil Corp."
    pub display_name: String,
    /// An email address to contact the account owner.
    pub email: String,
    /// Primary key for this record.
    pub id: Id,
    /// The catalog name for this account. Eg. "wiley" or "acmeCo"
    pub name: String,
    /// A case-folded, unicode normalized version of the catalog name. This is
    /// used to ensure that accounts have _functionally_ unique names, while
    /// always deferring to `name` for display purposes.
    pub unique_name: String,
    /// When this record was last updated.
    pub updated_at: DateTime<Utc>,
}

/// NewAccount represents the data required to insert a new Account, with
/// remaining fields from `Account` generated by Postgres.
#[derive(Debug, Deserialize, FromRow, Serialize)]
pub struct NewAccount {
    /// A freeform representation of the name. eg. "Wiley Coyote" or "Acme Anvil Corp."
    pub display_name: String,
    /// An email address to contact the account owner.
    pub email: String,
    /// A raw representation of the catalog name for this account. Eg. "wiley" or "acmeCo"
    pub name: String,
}
