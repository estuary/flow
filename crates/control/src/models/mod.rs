use std::convert::TryInto;
use std::fmt::Display;
use std::str::FromStr;

use base64::display::Base64Display;

pub mod accounts;
pub mod connector_images;
pub mod connectors;
pub mod credentials;

#[derive(Debug, thiserror::Error)]
pub enum MalformedIdError {
    #[error("id does not appear to be valid")]
    Decode,
    #[error("id with an unexpected length")]
    Length,
}

static ENCODING_CONFIG: base64::Config = base64::URL_SAFE_NO_PAD;

/// `Id` is a transparent wrapper around our non-sequential Postgres keys. We
/// wish to expose these Id values as base64 encoded values. This wrapper
/// handles all these serialization and deserialization concerns.
///
/// When creating a new Postgres-backed record or relationship, it is expected
/// to use `Id` for primary and foreign keys.
#[serde_as]
#[derive(
    Clone,
    Copy,
    Debug,
    Eq,
    Ord,
    PartialEq,
    PartialOrd,
    DeserializeFromStr,
    SerializeDisplay,
    sqlx::Type,
)]
#[sqlx(transparent)]
pub struct Id(i64);

impl Id {
    pub const fn new(inner: i64) -> Self {
        Self(inner)
    }

    /// Retrieves the raw value of the Id. Not to be directly exposed to
    /// customers, but not a secret.
    pub fn to_i64(&self) -> i64 {
        self.0
    }

    /// Converts the Id value to bytes. **Must** be the inverse operation of `from_bytes`.
    pub(crate) fn bytes(&self) -> [u8; 8] {
        self.0.to_be_bytes()
    }

    /// Converts bytes into an Id value. **Must** be the inverse operation of `bytes`.
    fn from_bytes(bytes: [u8; 8]) -> Self {
        Id::new(i64::from_be_bytes(bytes))
    }

    /// An Id that does not correspond to a persisted resource, but rather is a
    /// one-time use identifier. Nonces must not be reused.
    pub fn nonce() -> Self {
        Id::new(rand::random())
    }
}

impl Display for Id {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            Base64Display::with_config(&self.bytes(), ENCODING_CONFIG)
        )
    }
}

impl FromStr for Id {
    type Err = MalformedIdError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let bytes =
            base64::decode_config(s, ENCODING_CONFIG).map_err(|_| MalformedIdError::Decode)?;
        let bytes = bytes.try_into().map_err(|_| MalformedIdError::Length)?;
        Ok(Id::from_bytes(bytes))
    }
}

impl From<Id> for String {
    fn from(id: Id) -> Self {
        id.to_string()
    }
}
