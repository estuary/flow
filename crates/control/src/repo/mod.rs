pub mod accounts;
pub mod connector_images;
pub mod connectors;
pub mod credentials;
