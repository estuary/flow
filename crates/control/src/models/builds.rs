use chrono::{DateTime, Utc};
use models;
use serde::{Deserialize, Serialize};
use sqlx::{types::Json, FromRow};
use std::fmt;

use crate::models::accounts;
use crate::models::id::Id;

/// Status of the Build.
#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", tag = "type")]
pub enum State {
    Queued,
    Done,
    BuildFailed { code: Option<i32> },
    TestFailed { code: Option<i32> },
}

#[derive(Deserialize, FromRow, Serialize)]
pub struct Build {
    /// Account which owns this Build.
    pub account_id: Id<accounts::Account>,
    /// Root catalog built by this build, which may inline additional resources.
    /// The catalog may not be retrieved in all contexts.
    pub catalog: Option<Json<models::Catalog>>,
    /// When this record was created.
    pub created_at: DateTime<Utc>,
    /// Primary key for this record.
    pub id: Id<Build>,
    /// Connectors must be either a source or materialization.
    pub state: Json<State>,
    /// When this record was last updated.
    pub updated_at: DateTime<Utc>,
}

// TODO(johnny) impl Debug for Build is required by query_as!, but I'm not sure why.
impl fmt::Debug for Build {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Build")
            .field("account_id", &self.account_id)
            // catalog is omitted.
            .field("created_at", &self.created_at)
            .field("id", &self.id)
            .field("state", &self.state)
            .field("updated_at", &self.updated_at)
            .finish()
    }
}
