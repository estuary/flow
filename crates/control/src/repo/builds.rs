use crate::models::id::Id;
use crate::models::{
    accounts::Account,
    builds::{Build, State},
};
use futures::TryFutureExt;
use sqlx::{types::Json, PgPool, Postgres, Transaction};

pub async fn fetch_for_account<'e>(
    db: &PgPool,
    account_id: Id<Account>,
) -> Result<Vec<Build>, sqlx::Error> {
    sqlx::query_as!(
        Build,
        r#"
    SELECT
        account_id AS "account_id: Id<Account>",
        NULL AS "catalog: Json<models::Catalog>", -- Skip catalog JSON in index listing.
        created_at,
        id AS "id: Id<Build>",
        state AS "state: Json<State>",
        updated_at
    FROM builds
    WHERE account_id = $1
    "#,
        account_id as Id<Account>,
    )
    .fetch_all(db)
    .await
}

pub async fn fetch_one(
    db: &PgPool,
    build_id: Id<Build>,
    account_id: Id<Account>,
) -> Result<Build, sqlx::Error> {
    sqlx::query_as!(
        Build,
        r#"
    SELECT
        account_id AS "account_id: Id<Account>",
        catalog AS "catalog: Option<Json<models::Catalog>>",
        created_at,
        id AS "id: Id<Build>",
        state AS "state: Json<State>",
        updated_at
    FROM builds
    WHERE id = $1 AND account_id = $2
    "#,
        build_id as Id<Build>,
        account_id as Id<Account>,
    )
    .fetch_one(db)
    .await
}

pub async fn insert(
    db: &PgPool,
    catalog: models::Catalog,
    account_id: Id<Account>,
) -> Result<Build, sqlx::Error> {
    sqlx::query!(
        r#"
    INSERT INTO builds(account_id, catalog, state, created_at, updated_at)
    VALUES ($1, $2, $3, NOW(), NOW())
    RETURNING
        id as "id!: Id<Build>"
    "#,
        account_id as Id<Account>,
        Json(catalog) as Json<models::Catalog>,
        Json(State::Queued) as Json<State>,
    )
    .fetch_one(db)
    .and_then(|row| fetch_one(db, row.id, account_id))
    .await
}

// dequeue_build returns a queued Build to build, or None if the queue is empty.
// If a Some(Build) is returned, it's guaranteed that no parallel invocation of
// dequeue_build will return that same Build so long as the argument Transaction
// is alive.
//
// The query leverages PostgreSQL's advisory lock mechanism to obtain a
// transaction-scoped exclusive lock on the `id` of its returned row.
// `pg_try_advisory_xact_lock(id)` returns true if it obtains a lock on `id`,
// or false if it's locked already.
//
// It's therefore important that the query plan call `pg_try_advisory_xact_lock`
// on as few actual `id`'s as possible, as extra locking reduces parallelism.
// For this query, we're capitalizing on the index `builds_id_where_queued`.
// The optimizer uses it in an Index Scan to try successive increasing `id`'s
// until one is found which is lock-able.
//
// This index also means that the common case (there are no queued builds) is
// very cheap to poll, even if the total number of builds is large.
//
// If the query plan were to change on us then `pg_try_advisory_xact_lock(id)`
// could lock additional non-returned `id`'s, which would reduce the potential
// parallelism but would not be a correctness issue.
//
// All obtained locks are automatically released on transaction close.
// If a builder fails its lock is automatically removed, making the Build
// eligible again for dequeue by other workers.
//
/*
                                         QUERY PLAN
---------------------------------------------------------------------------------------------
Limit  (cost=0.13..6.18 rows=1 width=68)
->  Index Scan using builds_id_where_queued on builds  (cost=0.13..12.22 rows=2 width=68)
        Filter: pg_try_advisory_xact_lock(id)
*/
pub async fn dequeue_build<'c>(
    txn: &mut Transaction<'c, Postgres>,
) -> Result<Option<Build>, sqlx::Error> {
    sqlx::query_as!(
        Build,
        r#"
    SELECT
        account_id as "account_id: Id<Account>",
        catalog as "catalog: Option<Json<models::Catalog>>",
        created_at,
        id as "id: Id<Build>",
        state as "state: Json<State>",
        updated_at
    FROM builds
    WHERE state->>'type' = 'queued' AND
        pg_try_advisory_xact_lock(id)
    ORDER BY id ASC
    LIMIT 1
    "#,
    )
    .fetch_optional(txn)
    .await
}

pub async fn update_build_state<'c>(
    txn: &mut Transaction<'c, Postgres>,
    build_id: Id<Build>,
    state: State,
) -> Result<(), sqlx::Error> {
    let _ = sqlx::query!(
        r#"
    UPDATE builds SET
        state = $2,
        updated_at = NOW()
    WHERE id = $1
    RETURNING
        true AS "r"
    "#,
        build_id as Id<Build>,
        Json(state) as Json<State>,
    )
    .fetch_one(txn)
    .await?;
    Ok(())
}
