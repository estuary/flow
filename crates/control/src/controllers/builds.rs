use axum::extract::{Extension, Path};
use axum::response::IntoResponse;
use axum::Json;
use hyper::StatusCode;

use crate::context::AppContext;
use crate::error::AppError;
use crate::middleware::sessions::CurrentAccount;
use crate::models::{builds::Build, id::Id};
use crate::repo::builds as builds_repo;

pub mod routes;
mod view;

pub async fn index(
    Extension(ctx): Extension<AppContext>,
    CurrentAccount(account): CurrentAccount,
) -> Result<impl IntoResponse, AppError> {
    tracing::debug!(?account.id, "entered index");
    let builds = builds_repo::fetch_for_account(ctx.db(), account.id).await?;

    Ok((StatusCode::OK, view::index(builds)))
}

pub async fn create(
    Extension(ctx): Extension<AppContext>,
    CurrentAccount(account): CurrentAccount,
    Json(catalog): Json<models::Catalog>,
) -> Result<impl IntoResponse, AppError> {
    let build = builds_repo::insert(ctx.db(), catalog, account.id).await?;

    Ok((StatusCode::CREATED, view::create(build)))
}

pub async fn show(
    Extension(ctx): Extension<AppContext>,
    CurrentAccount(account): CurrentAccount,
    Path(build_id): Path<Id<Build>>,
) -> Result<impl IntoResponse, AppError> {
    let image = builds_repo::fetch_one(ctx.db(), build_id, account.id).await?;
    Ok((StatusCode::OK, view::show(image)))
}
