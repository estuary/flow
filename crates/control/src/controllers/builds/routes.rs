use crate::controllers::url_for;
use crate::models::{builds::Build, id::Id};

pub fn index() -> String {
    url_for("/builds")
}

pub fn show(build_id: Id<Build>) -> String {
    url_for(format!("/builds/{}", build_id.to_string()))
}
