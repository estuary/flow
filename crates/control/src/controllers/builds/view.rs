use axum::Json;

use crate::controllers::builds::routes;
use crate::controllers::json_api::{DocumentData, Links, Many, One, Resource};
use crate::models::builds::Build;

pub fn index(builds: Vec<Build>) -> Json<Many<Build>> {
    let resources = builds.into_iter().map(Resource::from).collect();
    let links = Links::default().put("self", routes::index());
    Json(DocumentData::new(resources, links))
}

pub fn create(build: Build) -> Json<One<Build>> {
    let resource = DocumentData::new(Resource::from(build), Links::default());
    Json(resource)
}

pub fn show(build: Build) -> Json<One<Build>> {
    let resource = DocumentData::new(Resource::from(build), Links::default());
    Json(resource)
}

impl From<Build> for Resource<Build> {
    fn from(build: Build) -> Self {
        let links = Links::default().put("self", routes::show(build.id));

        Resource {
            id: build.id,
            r#type: "build",
            attributes: build,
            links,
        }
    }
}
