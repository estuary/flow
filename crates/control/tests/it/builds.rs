use crate::support::{self, factory, redactor, test_context};

use control::models::builds::{Build, State};
use control::repo::builds::{dequeue_build, fetch_for_account, update_build_state};

#[tokio::test]
async fn index_test() {
    let mut t = test_context!();
    let batman = factory::BatmanAccount.create(t.db()).await;
    let joker = factory::JokerAccount.create(t.db()).await;
    let build_batman = factory::AcmeBuild.create(t.db(), batman.id).await;
    let build_joker = factory::AcmeBuild.create(t.db(), joker.id).await;

    let redactor = redactor::Redactor::default()
        .redact(batman.id, "account<batman>")
        .redact(joker.id, "account<joker>")
        .redact(build_batman.id, "build<batman>")
        .redact(build_joker.id, "build<joker>");

    // Batman views their build.
    t.login(batman.clone());
    let mut response = t.get("/builds").await;

    // Assert we get Batman's build (only).
    assert!(response.status().is_success());
    assert_json_snapshot!(redactor.response_json(&mut response).await.unwrap(), {
        ".data.*.attributes.created_at" => "[datetime]",
        ".data.*.attributes.updated_at" => "[datetime]",
    });
}

#[tokio::test]
async fn create_test() {
    let mut t = test_context!();
    let account = factory::BatmanAccount.create(t.db()).await;
    t.login(account.clone());
    let catalog = factory::AcmeBuild.attrs();

    let mut response = t.post("/builds", &catalog).await;

    // Expect build was created.
    let builds = fetch_for_account(t.db(), account.id)
        .await
        .expect("to insert test data");
    assert_eq!(1, builds.len());

    // Expect build was returned in the response.
    assert_eq!(201, response.status().as_u16());
    let redactor = redactor::Redactor::default()
        .redact(account.id, "a1")
        .redact(builds[0].id, "b1");
    assert_json_snapshot!(redactor.response_json(&mut response).await.unwrap(), {
        ".data.attributes.created_at" => "[datetime]",
        ".data.attributes.updated_at" => "[datetime]",
    });
}

#[tokio::test]
async fn show_test() {
    let mut t = test_context!();
    let batman = factory::BatmanAccount.create(t.db()).await;
    let joker = factory::JokerAccount.create(t.db()).await;
    let build = factory::AcmeBuild.create(t.db(), batman.id).await;

    let build_url = format!("/builds/{}", build.id);
    let redactor = redactor::Redactor::default()
        .redact(batman.id, "account<batman>")
        .redact(build.id, "build<batman>");

    // Joker attempts to view Batman's build, but it 404's.
    t.login(joker.clone());
    let response = t.get(&build_url).await;
    assert_eq!(response.status().canonical_reason(), Some("Not Found"));

    // Batman can view their own build.
    t.login(batman.clone());
    let mut response = t.get(&build_url).await;

    assert!(response.status().is_success());
    assert_json_snapshot!(redactor.response_json(&mut response).await.unwrap(), {
        ".data.attributes.created_at" => "[datetime]",
        ".data.attributes.updated_at" => "[datetime]",
    });
}

#[tokio::test]
async fn dequeue_builds_test() {
    let t = test_context!();
    let batman = factory::BatmanAccount.create(t.db()).await;
    let build_1 = factory::AcmeBuild.create(t.db(), batman.id).await;

    // First dequeue obtains build_1 and holds its transaction open.
    let mut txn_1 = t.db().begin().await.unwrap();
    let build = dequeue_build(&mut txn_1).await.unwrap();
    assert!(matches!(build, Some(Build { id, .. }) if id == build_1.id));

    // Second build is added, and a parallel transaction dequeues it.
    let build_2 = factory::AcmeBuild.create(t.db(), batman.id).await;

    let mut txn_2 = t.db().begin().await.unwrap();
    let build = dequeue_build(&mut txn_2).await.unwrap();
    assert!(matches!(build, Some(Build { id, .. }) if id == build_2.id));

    // Third dequeue is attempted, but doesn't obtain a build as both are locked.
    let mut txn_3 = t.db().begin().await.unwrap();
    let build = dequeue_build(&mut txn_3).await.unwrap();
    assert!(matches!(build, None));
    std::mem::drop(txn_3);

    // Second transaction updates its build and commits.
    update_build_state(&mut txn_2, build_2.id, State::Done)
        .await
        .unwrap();
    txn_2.commit().await.unwrap();

    // First transaction aborts.
    std::mem::drop(txn_1);

    // First build is now eligible for dequeue again.
    let mut txn_4 = t.db().begin().await.unwrap();
    let build = dequeue_build(&mut txn_4).await.unwrap();
    assert!(matches!(build, Some(Build { id, .. }) if id == build_1.id));

    // It's completed.
    update_build_state(&mut txn_4, build_1.id, State::TestFailed { code: None })
        .await
        .unwrap();
    txn_4.commit().await.unwrap();

    // No further builds are queued.
    let mut txn_5 = t.db().begin().await.unwrap();
    let build = dequeue_build(&mut txn_5).await.unwrap();
    assert!(matches!(build, None));
}
