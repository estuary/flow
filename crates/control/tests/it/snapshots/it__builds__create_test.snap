---
source: crates/control/tests/it/builds.rs
assertion_line: 52
expression: redactor.response_json(&mut response).await.unwrap()
---
{
  "data": {
    "attributes": {
      "account_id": "[a1]",
      "catalog": {
        "collections": {
          "acmeCo/collection": {
            "key": [
              "/key"
            ],
            "schema": {
              "properties": {
                "key": {
                  "type": "integer"
                }
              },
              "required": "key",
              "type": "object"
            }
          }
        }
      },
      "created_at": "[datetime]",
      "id": "[b1]",
      "state": {
        "type": "queued"
      },
      "updated_at": "[datetime]"
    },
    "id": "[b1]",
    "links": {
      "self": "http://127.0.0.1:0/builds/[b1]"
    },
    "type": "build"
  }
}
